CovertInfiltration/Src/CovertInfiltration/Classes/UIListener_CovertActionCompleted.uc

```diff
@@ -0,0 +1,136 @@
+//---------------------------------------------------------------------------------------
+//  AUTHOR:  statusNone
+//  PURPOSE: This class is used to track when a covert action is completed in order to
+//	apply soldier will loss
+//---------------------------------------------------------------------------------------
+//  WOTCStrategyOverhaul Team
+//---------------------------------------------------------------------------------------
+
+class UIListener_CovertActionCompleted extends UIScreenListener config(Infiltration);
+
+//values from config represent a percentage to be removed from total will e.g.(0.25 = 25%, 0.50 = 50%)
+var config float MIN_WILL_LOSS;
+var config float MAX_WILL_LOSS;
+
+event OnInit(UIScreen Screen)
+{
+	local XComGameState_CovertAction CovertAction;
+	local UICovertActionReport CovertActionReport;
+
+	CovertActionReport = UICovertActionReport(Screen);
+	if (CovertActionReport == none) return;
+
+	CovertAction = XComGameState_CovertAction(`XCOMHISTORY.GetGameStateForObjectID(CovertActionReport.ActionRef.ObjectID));
+
+	ApplyWillLossToSoldiers(CovertAction, CovertActionReport);
+}
+
+function ApplyWillLossToSoldiers(XComGameState_CovertAction CovertAction, UICovertActionReport CovertActionReport)
+{
+	local XComGameState NewGameState;
+	local XComGameState_StaffSlot StaffSlotState;
+	local XComGameState_Unit UnitState;
+
+	local int idx;
+
+	NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: Applying will loss to soldiers");
+
+	for (idx = 0; idx < CovertAction.StaffSlots.Length; idx++)
+	{
+		StaffSlotState = CovertAction.GetStaffSlot(idx);
+		UnitState = XComGameState_Unit(NewGameState.ModifyStateObject(class'XComGameState_Unit', StaffSlotState.GetAssignedStaff().ObjectID));
+
+		if (StaffSlotState.IsSlotFilled() && UnitState.IsSoldier() && !UnitState.IsInjured() && !UnitState.bCaptured)
+		{
+			UnitState.SetCurrentStat(eStat_Will, GetWillLoss(UnitState));
+			UpdateWillRecovery(NewGameState, UnitState);
+			UnitState.UpdateMentalState();
+			ShowTiredOnReport(CovertActionReport, StaffSlotState, UnitState, idx);
+		}
+	}
+
+	`XCOMGAME.GameRuleset.SubmitGameState(NewGameState);
+}
+
+	function int GetWillLoss(XComGameState_Unit UnitState)
+{
+	local int WillToLose, LowestWill;
+
+	WillToLose = int(UnitState.GetMaxStat(eStat_Will) * RandRange(MIN_WILL_LOSS, MAX_WILL_LOSS));
```

Comment:
Need to use macro:
* `SYNC_RAND(max)` - integer between 0 and `max`
* `SYNC_FRAND` - float between 0 and 1

There is no automatic range so you will need to get creative ;)